package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swift-nav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <libgen.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/queue.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <libpiksi/logging.h>
+
+#include "path_validator.h"
+#include "fio_debug.h"
+
+static bool validate_path(const char* basedir_path, const char* path);
+
+typedef struct path_node {
+  char path[PATH_MAX];
+  LIST_ENTRY(path_node) entries;
+} path_node_t;
+
+typedef LIST_HEAD(path_nodes_head, path_node) path_nodes_head_t;
+
+struct path_validator_s
+{
+  char base_paths[4096];
+  path_nodes_head_t path_list;
+  size_t allowed_count;
+};
+
+path_validator_t *path_validator_create(void)
+{
+  path_validator_t *ctx = malloc(sizeof(path_validator_t));
+
+  if (ctx == NULL)
+    return NULL;
+
+  LIST_INIT(&(ctx->path_list));
+  ctx->allowed_count = 0;
+
+  return ctx;
+}
+
+void path_validator_destroy(path_validator_t **pctx)
+{
+  if (*pctx == NULL)
+    return;
+
+  path_node_t *node;
+
+  while (!LIST_EMPTY(&((*pctx)->path_list))) {
+    node = LIST_FIRST(&((*pctx)->path_list));
+    LIST_REMOVE(node, entries);
+    free(node);
+  }
+
+  free(*pctx);
+  *pctx = NULL;
+}
+
+bool path_validator_check(path_validator_t *ctx, const char* path)
+{
+  path_node_t *node;
+
+  LIST_FOREACH(node, &ctx->path_list, entries) {
+    if(validate_path(node->path, path))
+      return true;
+  }
+
+  return false;
+}
+
+void path_validator_allow_path(path_validator_t *ctx, const char* path)
+{
+  assert( path != NULL );
+
+  if (strlen(path) == 0) {
+
+    fprintf(stderr, "ERROR: directories allowed for SBP fileio must not be empty.\n");
+    exit(EXIT_FAILURE);
+  }
+
+  path_node_t *node = malloc(sizeof(path_node_t));
+  node->path[0] = '\0';
+
+  strncpy(node->path, path, sizeof(node->path));
```

Comment:
we should make sure we check lengths here, `strncpy` is known to do some funny stuff: https://stackoverflow.com/a/1258577

maybe even add a test to cover path overflow?

Reply:
Added a test and removed strncpy

Reply:
Oops, didn't actually remove this strncpy...

Reply:
Ok, removed it for real now, and added an overflow test for this function.